src/bit-fields-macros/src/builder.rs

```diff
@@ -0,0 +1,830 @@
+                ///
+                /// This only considers defined bit flags.
+                pub fn superset(&self, other: &Self) -> bool {
+                    #superset
+                }
+                /// Returns `true` if `self` is a [`subset`](https://en.wikipedia.org/wiki/Subset) of `other`.
+                ///
+                /// This only considers defined bit flags.
+                pub fn subset(&self, other: &Self) -> bool {
+                    #subset
+                }
+                /// Returns `true` if `self` and `other` are [`disjoint sets`](https://en.wikipedia.org/wiki/Disjoint_sets).
+                ///
+                /// This only considers defined bit flags.
+                pub fn disjoint(&self, other: &Self) -> bool {
+                    #disjoint
+                }
+                #intersection
+                #union_fn
+            }
+        };
+
+        let bit_index = (0..self.data_type.size())
+            .map(|i| {
+                quote! {
+                    impl bit_fields::BitIndex<#data_type,#i> for #struct_name {
+                        fn bit(&self) -> bit_fields::Bit<#data_type,#i> {
+                            bit_fields::Bit(&self.0)
+                        }
+                    }
+                    impl bit_fields::BitIndexMut<#data_type,#i> for #struct_name {
+                        fn bit_mut(&mut self) -> bit_fields::BitMut<#data_type,#i> {
+                            bit_fields::BitMut(&mut self.0)
+                        }
+                    }
+                }
+            })
+            .collect::<TokenStream>();
+
+        let index_fn = quote! {
+            /// Returns a reference to the `N`th bit.
+            #[allow(clippy::same_name_method)]
+            pub fn bit<const N: u8>(&self) -> bit_fields::Bit<#data_type,N>
+            where
+                Self: bit_fields::BitIndex<#data_type,N>,
+            {
+                <Self as bit_fields::BitIndex<#data_type,N>>::bit(self)
+            }
+            /// Returns a mutable reference to the `N`th bit.
+            #[allow(clippy::same_name_method)]
+            pub fn bit_mut<const N: u8>(&mut self) -> bit_fields::BitMut<#data_type,N>
+            where
+                Self: bit_fields::BitIndexMut<#data_type,N>,
+            {
+                <Self as bit_fields::BitIndexMut<#data_type,N>>::bit_mut(self)
+            }
+        };
+
+        // - Constructing a bit field from a set of bit flags
+        let from_type_parts = |err: TokenStream| {
+            if data_type.is_nonzero() {
+                (
+                    quote! { Option<#err> },
+                    quote! {
+                        match value {
+                            0 => Err(None),
+                            _ => Ok(unsafe { Self(#data_type::new_unchecked(0)) })
+                        }
+                    },
```

Comment:
As per offline discussion, we can turn `TryFromFlagSetError` into an enum so that the error type is no longer an Option. That way we can get rid of this `unsafe`:

```suggestion
                    quote! { #err },
                    quote! {
                        #data_type::new(value)
                            .map(Self)
                            .ok_or(TryFromFlagSetError::NoFlagsSet)
                    },
```

Reply:
Fixed, renamed a few error types to fit.